### PR TITLE
Avoid using masked quote IDs in the abandon cart reclaim URL

### PR DESCRIPTION
**Pull request by liamoneill**
A HMAC is used instead of Magento's masked quote IDs because every cart doesn't necessary have a masked ID and creating a new masked ID can fail (via a foreign key constraint) if the cart has been deleted by the time the cart event has been picked up by the cron job.

This PR adds a new configuration "HMAC Secret" field to the Admin UI which needs to be set in order to set a quote reclaim & validate and parse the token.

### CODE REVIEW

File: Helper/ReclaimCartTokenHelper.php [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    public function parseAndValidateToken(string $token, string $secret): ?string
+    {
+        try {
+            // URL decode the data before base64 decoding it just in case it has been URL
+            //      encoded twice as it passed through different external services.
+            $data = json_decode(base64_decode(urldecode($token)), true);
+
+            $quoteId = $data['id'];
+            $salt = $data['sa'];
+            $timestamp = $data['ts'];
```
Comment (dylan-chong) [inline Helper/ReclaimCartTokenHelper.php:68]: Needs comment/renaming about what the purpose of the salt/timestamp is

File: Model/Event/Transport/Adapter/GraphQL/PayloadConverter.php [hunk 2]
```diff
@@ -872,16 +882,18 @@ private function getReclaimCartUrl(array $quote, array $area): ?string
...
-            /** @var $quoteIdMask \Magento\Quote\Model\QuoteIdMask */
-            $quoteIdMask = $this->quoteIdMaskFactory->create()->load($quoteId, 'quote_id');
-            if ($quoteIdMask->getMaskedId() === null) {
-                $this->logger->debug('Masked quote id did not exist, creating a new masked id', ['quoteId' => $quoteId]);
-                $quoteIdMask->setQuoteId($quoteId)->save();
+            $now = new \DateTime();
+            $secret = $this->config->getHmacSecret();
+
+            if (empty($secret)) {
+                return null;
```
Comment (dylan-chong) [inline Model/Event/Transport/Adapter/GraphQL/PayloadConverter.php:889]: might be worth printing a warning here?
or logging an error

File: etc/adminhtml/system.xml [hunk 3]
```diff
@@ -29,6 +29,13 @@
                         <field id="enabled">1</field>
                     </depends>
                 </field>
+                <field id="hmac_secret" translate="label" type="text" sortOrder="30" showInDefault="1" showInWebsite="1" showInStore="1">
+                    <label>Abandoned Carts HMAC secret</label>
```
Comment (dylan-chong) [inline etc/adminhtml/system.xml:33]: should we also say how to generate it here, and what length it should be

Comment (liamoneill) [reply etc/adminhtml/system.xml:33]: We will be managing these settings for the first couple of customers so let's document it in a wiki page for now.

Longer term I'd like to see if we can get one of the extension's setup scripts to generate it if it doesn't already exist. We could also expose this value through an admin endpoint (which requires admin auth token) and therefore we can programatically query for it.

File: Model/Event/Transport/Adapter/GraphQL/PayloadConverter.php (hunk 2 above)
Comment (liamoneill) [reply Model/Event/Transport/Adapter/GraphQL/PayloadConverter.php:889]: Good call, added a warning error here.

File: Helper/ReclaimCartTokenHelper.php (hunk 1 above)
Comment (liamoneill) [reply Helper/ReclaimCartTokenHelper.php:72]: 👍 good call out. Added a comment inside the generation function to explain their purpose.

**Review by dylan-chong — APPROVED** [review]
Works great with the new hmac code!
Only issues are the existing ones
- if you pass in an invalid hmac to the reclaim endpoint it takes a minute to redirect you to the home page
- if you are not signed in, and the cart is an identified cart, you are not asked to sign in, you are just shown an empty cart. not sure if you have checked what happens on bobux's klaviyo emails because the bobux session timeout is 1 day IIRC

**Comment by liamoneill** [comment]
> * if you pass in an invalid hmac to the reclaim endpoint it takes a minute to redirect you to the home page


Re ~1 minute redirect. That's not something I've been able to reproduce and is likely an issue with the performance of the virtualbox dev environment. @dylan-chong mentioned that it didn't occur consistently and also occurred with carts that were successfully reclaimed.

> * if you are not signed in, and the cart is an identified cart, you are not asked to sign in, you are just shown an empty cart. not sure if you have checked what happens on bobux's klaviyo emails because the bobux session timeout is 1 day IIRC


That's a great catch, thanks Dylan! This is something we should investigate if we can improve in a future PR.

**Pull request merged**